File: backend/app/schemas/budget.py
from datetime import date

from pydantic import BaseModel, Field


class BudgetProfileCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    description: str | None = None


class BudgetProfileUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=120)
    description: str | None = None


class BudgetProfileRead(BaseModel):
    id: int
    name: str
    description: str | None


class BudgetCategoryRead(BaseModel):
    id: int
    profile_id: int
    name: str
    sort_order: int
    color: str


class BudgetCategoryCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    color: str = "#64748b"


class BudgetCategoryUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=120)
    color: str | None = None
    # 'all' altera o catálogo (todos os meses); 'from_month' aplica override do year_month em diante.
    scope: str = "all"
    year_month: str | None = None


class BudgetTagCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    color: str = "#22c55e"


class BudgetTagUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=120)
    color: str | None = None


class BudgetTagRead(BaseModel):
    id: int
    profile_id: int
    name: str
    color: str
    usage_count: int = 0


class BudgetIncomeSourceCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    recurrence_hint: str = "variable"


class BudgetIncomeSourceUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=120)
    recurrence_hint: str | None = None
    is_active: bool | None = None


class BudgetIncomeSourceRead(BaseModel):
    id: int
    profile_id: int
    name: str
    recurrence_hint: str
    sort_order: int
    is_active: bool


class BudgetMonthIncomeItem(BaseModel):
    id: int | None = None
    source_id: int | None = None
    label: str
    amount_brl: float
    recurring: bool = False
    received: bool = False


class BudgetIncomeEntryCreate(BaseModel):
    label: str = Field(min_length=1, max_length=120)
    amount_brl: float = Field(gt=0)
    recurring_12_months: bool = False


class BudgetIncomeEntryUpdate(BaseModel):
    label: str | None = Field(default=None, min_length=1, max_length=120)
    amount_brl: float | None = Field(default=None, gt=0)
    received: bool | None = None


class BudgetMonthIncomesUpdate(BaseModel):
    items: list[BudgetMonthIncomeItem]


class BudgetMonthTargetItem(BaseModel):
    category_id: int
    percent: float
    name: str | None = None
    color: str | None = None


class BudgetMonthTargetsUpdate(BaseModel):
    planned_income_brl: float | None = None
    targets: list[BudgetMonthTargetItem]
    # Categorias novas a incluir (com 0%) nos meses seguintes que já tiverem metas próprias.
    propagate_category_ids: list[int] = Field(default_factory=list)
    # Copia o conjunto completo (categorias + %) para meses seguintes com metas próprias.
    apply_to_following_months: bool = False


class BudgetRemoveTargetCategories(BaseModel):
    """Remove categorias dos conjuntos de metas (sem exigir soma 100%)."""

    category_ids: list[int] = Field(min_length=1)
    # Se true, remove também dos meses > year_month que já tiverem metas próprias.
    apply_to_following_months: bool = False
    # Se true, remove do mês year_month (quando ele já tiver linhas próprias).
    apply_to_current: bool = True


class BudgetMonthPatch(BaseModel):
    planned_income_brl: float | None = None


class BudgetExpenseEntryCreate(BaseModel):
    description: str = Field(min_length=1, max_length=240)
    event_date: str
    amount_brl: float = Field(gt=0)
    category_id: int
    tag_id: int | None = None
    recurring: bool = False
    indefinite: bool = False
    end_year_month: str | None = None


class BudgetExpenseEntryUpdate(BaseModel):
    description: str | None = Field(default=None, min_length=1, max_length=240)
    event_date: str | None = None
    amount_brl: float | None = Field(default=None, gt=0)
    category_id: int | None = None
    tag_id: int | None = None
    indefinite: bool | None = None
    end_year_month: str | None = None


class BudgetRecurringExpenseRead(BaseModel):
    id: int
    profile_id: int
    description: str
    amount_brl: float
    category_id: int
    category_name: str | None = None
    tag_id: int | None
    tag_name: str | None = None
    day_of_month: int
    start_year_month: str
    end_year_month: str | None
    indefinite: bool
    is_active: bool


class BudgetCategoryUsageSummary(BaseModel):
    id: int
    profile_id: int
    name: str
    sort_order: int
    color: str
    transaction_count: int
    recurring_count: int
    can_delete: bool


class BudgetCategoryUsageTransaction(BaseModel):
    id: int
    event_date: str
    year_month: str
    description: str
    amount_brl: float
    recurring: bool


class BudgetCategoryUsageDetail(BudgetCategoryUsageSummary):
    transactions: list[BudgetCategoryUsageTransaction] = Field(default_factory=list)
    recurring_expenses: list[BudgetRecurringExpenseRead] = Field(default_factory=list)


class BudgetTransactionCreate(BaseModel):
    transaction_type: str
    event_date: str
    description: str = Field(min_length=1, max_length=240)
    amount_brl: float = Field(gt=0)
    category_id: int | None = None
    tag_id: int | None = None
    income_source_id: int | None = None
    notes: str | None = None


class BudgetTransactionUpdate(BaseModel):
    event_date: str | None = None
    description: str | None = Field(default=None, min_length=1, max_length=240)
    amount_brl: float | None = Field(default=None, gt=0)
    category_id: int | None = None
    tag_id: int | None = None
    income_source_id: int | None = None
    notes: str | None = None
    settled: bool | None = None


class BudgetTransactionRead(BaseModel):
    id: int
    profile_id: int
    month_id: int
    transaction_type: str
    event_date: str
    description: str
    amount_brl: float
    category_id: int | None
    category_name: str | None = None
    tag_id: int | None
    tag_name: str | None = None
    tag_color: str | None = None
    income_source_id: int | None
    notes: str | None
    recurring: bool = False
    recurring_expense_id: int | None = None
    settled: bool = False


class BudgetCategoryKpiRead(BaseModel):
    category_id: int
    category_name: str
    color: str
    percent: float
    target_brl: float
    spent_brl: float
    remaining_brl: float
    usage_percent: float
    exceeded: bool
    transaction_count: int


class BudgetMonthSnapshotRead(BaseModel):
    profile_id: int
    year_month: str
    planned_income_brl: float | None
    income_total_brl: float
    expense_total_brl: float
    remaining_brl: float
    income_usage_percent: float
    categories: list[BudgetCategoryKpiRead]
    incomes: list[BudgetMonthIncomeItem]
    transactions: list[BudgetTransactionRead]
    # True quando as metas do mês são herdadas do mês anterior (o mês não tem metas próprias).
    targets_inherited: bool = False


class DashboardMonthRow(BaseModel):
    year_month: str
    income_brl: float
    expense_brl: float


class DashboardSliceRead(BaseModel):
    id: int
    name: str
    color: str
    amount_brl: float
    percent: float


class BudgetDashboardRead(BaseModel):
    profile_id: int
    months: int
    focus_year_month: str
    forward_months: int
    from_year_month: str | None = None
    to_year_month: str | None = None
    result_brl: float
    income_brl: float
    expense_brl: float
    balance_brl: float
    timeline: list[DashboardMonthRow]
    expenses_by_tag: list[DashboardSliceRead]
    expenses_by_category: list[DashboardSliceRead]
    incomes_by_tag: list[DashboardSliceRead] = []


class ActiveBudgetProfileRead(BaseModel):
    profile_id: int | None


class ActiveBudgetProfileUpdate(BaseModel):
    profile_id: int | None

File: frontend/src/lib/api/budget.ts
import { API_BASE_URL } from './config';

async function parseResponse<T>(response: Response): Promise<T> {
  if (!response.ok) {
    throw new Error(await response.text());
  }
  if (response.status === 204) {
    return undefined as T;
  }
  return response.json() as Promise<T>;
}

export type BudgetProfile = {
  id: number;
  name: string;
  description: string | null;
};

export type BudgetTag = {
  id: number;
  profile_id: number;
  name: string;
  color: string;
  usage_count: number;
};

export type BudgetCategory = {
  id: number;
  profile_id: number;
  name: string;
  sort_order: number;
  color: string;
};

export type BudgetCategoryUsageSummary = BudgetCategory & {
  transaction_count: number;
  recurring_count: number;
  can_delete: boolean;
};

export type BudgetCategoryUsageTransaction = {
  id: number;
  event_date: string;
  year_month: string;
  description: string;
  amount_brl: number;
  recurring: boolean;
};

export type BudgetCategoryUsageDetail = BudgetCategoryUsageSummary & {
  transactions: BudgetCategoryUsageTransaction[];
  recurring_expenses: BudgetRecurringExpense[];
};

export type BudgetCategoryScope = 'all' | 'from_month';

export type BudgetIncomeSource = {
  id: number;
  profile_id: number;
  name: string;
  recurrence_hint: string;
  sort_order: number;
  is_active: boolean;
};

export type BudgetMonthIncomeItem = {
  id?: number | null;
  source_id?: number | null;
  label: string;
  amount_brl: number;
  recurring?: boolean;
  received?: boolean;
};

export type BudgetCategoryKpi = {
  category_id: number;
  category_name: string;
  color: string;
  percent: number;
  target_brl: number;
  spent_brl: number;
  remaining_brl: number;
  usage_percent: number;
  exceeded: boolean;
  transaction_count: number;
};

export type BudgetTransaction = {
  id: number;
  profile_id: number;
  month_id: number;
  transaction_type: 'income' | 'expense';
  event_date: string;
  description: string;
  amount_brl: number;
  category_id: number | null;
  category_name: string | null;
  tag_id: number | null;
  tag_name: string | null;
  tag_color: string | null;
  income_source_id: number | null;
  notes: string | null;
  recurring?: boolean;
  recurring_expense_id?: number | null;
  settled?: boolean;
};

export type BudgetRecurringExpense = {
  id: number;
  profile_id: number;
  description: string;
  amount_brl: number;
  category_id: number;
  category_name: string | null;
  tag_id: number | null;
  tag_name: string | null;
  day_of_month: number;
  start_year_month: string;
  end_year_month: string | null;
  indefinite: boolean;
  is_active: boolean;
};

export type BudgetMonthSnapshot = {
  profile_id: number;
  year_month: string;
  planned_income_brl: number | null;
  income_total_brl: number;
  expense_total_brl: number;
  remaining_brl: number;
  income_usage_percent: number;
  categories: BudgetCategoryKpi[];
  incomes: BudgetMonthIncomeItem[];
  transactions: BudgetTransaction[];
  targets_inherited: boolean;
};

export type DashboardSlice = {
  id: number;
  name: string;
  color: string;
  amount_brl: number;
  percent: number;
};

export type BudgetDashboard = {
  profile_id: number;
  months: number;
  focus_year_month: string;
  forward_months: number;
  from_year_month: string | null;
  to_year_month: string | null;
  result_brl: number;
  income_brl: number;
  expense_brl: number;
  balance_brl: number;
  timeline: { year_month: string; income_brl: number; expense_brl: number }[];
  expenses_by_tag: DashboardSlice[];
  expenses_by_category: DashboardSlice[];
  incomes_by_tag: DashboardSlice[];
};

export async function listBudgetProfiles(fetcher: typeof fetch = fetch): Promise<BudgetProfile[]> {
  const response = await fetcher(`${API_BASE_URL}/budget/profiles`);
  return parseResponse<BudgetProfile[]>(response);
}

export async function createBudgetProfile(
  payload: { name: string; description?: string | null },
  fetcher: typeof fetch = fetch
): Promise<BudgetProfile> {
  const response = await fetcher(`${API_BASE_URL}/budget/profiles`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload)
  });
  return parseResponse<BudgetProfile>(response);
}

export async function updateBudgetProfile(
  id: number,
  payload: { name?: string; description?: string | null },
  fetcher: typeof fetch = fetch
): Promise<BudgetProfile> {
  const response = await fetcher(`${API_BASE_URL}/budget/profiles/${id}`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload)
  });
  return parseResponse<BudgetProfile>(response);
}

export async function deleteBudgetProfile(id: number, fetcher: typeof fetch = fetch): Promise<void> {
  const response = await fetcher(`${API_BASE_URL}/budget/profiles/${id}`, { method: 'DELETE' });
  await parseResponse<void>(response);
}

export async function getActiveBudgetProfileId(fetcher: typeof fetch = fetch): Promise<number | null> {
  const response = await fetcher(`${API_BASE_URL}/budget/active`);
  const body = await parseResponse<{ profile_id: number | null }>(response);
  return body.profile_id;
}

export async function setActiveBudgetProfileId(
  profileId: number | null,
  fetcher: typeof fetch = fetch
): Promise<number | null> {
  const response = await fetcher(`${API_BASE_URL}/budget/active`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ profile_id: profileId })
  });
  const body = await parseResponse<{ profile_id: number | null }>(response);
  return body.profile_id;
}

export type BudgetSnapshotView = 'full' | 'targets' | 'incomes' | 'expenses' | 'settlement';

export async function getMonthSnapshot(
  profileId: number,
  yearMonth: string,
  view: BudgetSnapshotView = 'full',
  fetcher: typeof fetch = fetch
): Promise<BudgetMonthSnapshot> {
  const query = view === 'full' ? '' : `?view=${encodeURIComponent(view)}`;
  const response = await fetcher(
    `${API_BASE_URL}/budget/profiles/${profileId}/months/${yearMonth}${query}`
  );
  return parseResponse<BudgetMonthSnapshot>(response);
}

export async function updateMonthTargets(
  profileId: number,
  yearMonth: string,
  payload: {
    planned_income_brl?: number | null;
    targets: { category_id: number; percent: number; name?: string; color?: string }[];
    /** Inclui estas categorias (0%) nos meses seguintes que já tiverem metas próprias. */
    propagate_category_ids?: number[];
    /** Copia o conjunto completo (categorias + %) para os meses seguintes customizados. */
    apply_to_following_months?: boolean;
  },
  fetcher: typeof fetch = fetch
): Promise<BudgetMonthSnapshot> {
  const response = await fetcher(`${API_BASE_URL}/budget/profiles/${profileId}/months/${yearMonth}/targets`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload)
  });
  return parseResponse<BudgetMonthSnapshot>(response);
}

export async function removeTargetCategories(
  profileId: number,
  yearMonth: string,
  payload: {
    category_ids: number[];
    apply_to_current?: boolean;
    apply_to_following_months?: boolean;
  },
  fetcher: typeof fetch = fetch
): Promise<BudgetMonthSnapshot> {
  const response = await fetcher(
    `${API_BASE_URL}/budget/profiles/${profileId}/months/${yearMonth}/targets/remove-categories`,
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload)
    }
  );
  return parseResponse<BudgetMonthSnapshot>(response);
}

export async function patchMonth(
  profileId: number,
  yearMonth: string,
  payload: { planned_income_brl?: number | null },
  fetcher: typeof fetch = fetch
): Promise<BudgetMonthSnapshot> {
  const response = await fetcher(`${API_BASE_URL}/budget/profiles/${profileId}/months/${yearMonth}`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload)
  });
  return parseResponse<BudgetMonthSnapshot>(response);
}

export async function updateMonthIncomes(
  profileId: number,
  yearMonth: string,
  items: BudgetMonthIncomeItem[],
  fetcher: typeof fetch = fetch
): Promise<BudgetMonthSnapshot> {
  const response = await fetcher(`${API_BASE_URL}/budget/profiles/${profileId}/months/${yearMonth}/incomes`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ items })
  });
  return parseResponse<BudgetMonthSnapshot>(response);
}

export async function createMonthIncome(
  profileId: number,
  yearMonth: string,
  payload: { label: string; amount_brl: number; recurring_12_months?: boolean },
  fetcher: typeof fetch = fetch
): Promise<BudgetMonthSnapshot> {
  const response = await fetcher(`${API_BASE_URL}/budget/profiles/${profileId}/months/${yearMonth}/incomes`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload)
  });
  return parseResponse<BudgetMonthSnapshot>(response);
}

export async function updateMonthIncome(
  profileId: number,
  yearMonth: string,
  incomeId: number,
  payload: { label?: string; amount_brl?: number; received?: boolean },
  fetcher: typeof fetch = fetch
): Promise<BudgetMonthSnapshot> {
  const response = await fetcher(
    `${API_BASE_URL}/budget/profiles/${profileId}/months/${yearMonth}/incomes/${incomeId}`,
    {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload)
    }
  );
  return parseResponse<BudgetMonthSnapshot>(response);
}

export async function deleteMonthIncome(
  profileId: number,
  yearMonth: string,
  incomeId: number,
  fetcher: typeof fetch = fetch
): Promise<BudgetMonthSnapshot> {
  const response = await fetcher(
    `${API_BASE_URL}/budget/profiles/${profileId}/months/${yearMonth}/incomes/${incomeId}`,
    { method: 'DELETE' }
  );
  return parseResponse<BudgetMonthSnapshot>(response);
}

export async function copyPreviousMonthIncomes(
  profileId: number,
  yearMonth: string,
  fetcher: typeof fetch = fetch
): Promise<BudgetMonthSnapshot> {
  const response = await fetcher(
    `${API_BASE_URL}/budget/profiles/${profileId}/months/${yearMonth}/copy-previous-incomes`,
    { method: 'POST' }
  );
  return parseResponse<BudgetMonthSnapshot>(response);
}

export async function createTransaction(
  profileId: number,
  yearMonth: string,
  payload: {
    transaction_type: 'income' | 'expense';
    event_date: string;
    description: string;
    amount_brl: number;
    category_id?: number | null;
    tag_id?: number | null;
    income_source_id?: number | null;
    notes?: string | null;
  },
  fetcher: typeof fetch = fetch
): Promise<BudgetTransaction> {
  const response = await fetcher(
    `${API_BASE_URL}/budget/profiles/${profileId}/months/${yearMonth}/transactions`,
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload)
    }
  );
  return parseResponse<BudgetTransaction>(response);
}

export async function updateTransaction(
  profileId: number,
  transactionId: number,
  payload: {
    event_date?: string;
    description?: string;
    amount_brl?: number;
    category_id?: number | null;
    tag_id?: number | null;
    income_source_id?: number | null;
    notes?: string | null;
    settled?: boolean;
  },
  fetcher: typeof fetch = fetch
): Promise<BudgetTransaction> {
  const response = await fetcher(
    `${API_BASE_URL}/budget/profiles/${profileId}/transactions/${transactionId}`,
    {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload)
    }
  );
  return parseResponse<BudgetTransaction>(response);
}

export async function deleteTransaction(
  profileId: number,
  transactionId: number,
  fetcher: typeof fetch = fetch
): Promise<void> {
  const response = await fetcher(
    `${API_BASE_URL}/budget/profiles/${profileId}/transactions/${transactionId}`,
    { method: 'DELETE' }
  );
  await parseResponse<void>(response);
}

export async function listTags(profileId: number, fetcher: typeof fetch = fetch): Promise<BudgetTag[]> {
  const response = await fetcher(`${API_BASE_URL}/budget/profiles/${profileId}/tags`);
  return parseResponse<BudgetTag[]>(response);
}

export async function createTag(
  profileId: number,
  payload: { name: string; color: string },
  fetcher: typeof fetch = fetch
): Promise<BudgetTag> {
  const response = await fetcher(`${API_BASE_URL}/budget/profiles/${profileId}/tags`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload)
  });
  return parseResponse<BudgetTag>(response);
}

export async function updateTag(
  profileId: number,
  tagId: number,
  payload: { name?: string; color?: string },
  fetcher: typeof fetch = fetch
): Promise<BudgetTag> {
  const response = await fetcher(`${API_BASE_URL}/budget/profiles/${profileId}/tags/${tagId}`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload)
  });
  return parseResponse<BudgetTag>(response);
}

export async function deleteTag(
  profileId: number,
  tagId: number,
  fetcher: typeof fetch = fetch
): Promise<void> {
  const response = await fetcher(`${API_BASE_URL}/budget/profiles/${profileId}/tags/${tagId}`, {
    method: 'DELETE'
  });
  await parseResponse<void>(response);
}

export async function listCategories(
  profileId: number,
  fetcher: typeof fetch = fetch
): Promise<BudgetCategory[]> {
  const response = await fetcher(`${API_BASE_URL}/budget/profiles/${profileId}/categories`);
  return parseResponse<BudgetCategory[]>(response);
}

export async function listCategoriesUsage(
  profileId: number,
  fetcher: typeof fetch = fetch
): Promise<BudgetCategoryUsageSummary[]> {
  const response = await fetcher(`${API_BASE_URL}/budget/profiles/${profileId}/categories/usage`);
  return parseResponse<BudgetCategoryUsageSummary[]>(response);
}

export async function getCategoryUsage(
  profileId: number,
  categoryId: number,
  fetcher: typeof fetch = fetch
): Promise<BudgetCategoryUsageDetail> {
  const response = await fetcher(
    `${API_BASE_URL}/budget/profiles/${profileId}/categories/${categoryId}/usage`
  );
  return parseResponse<BudgetCategoryUsageDetail>(response);
}

export async function deleteCategoryExpenses(
  profileId: number,
  categoryId: number,
  fetcher: typeof fetch = fetch
): Promise<BudgetCategoryUsageDetail> {
  const response = await fetcher(
    `${API_BASE_URL}/budget/profiles/${profileId}/categories/${categoryId}/expenses`,
    { method: 'DELETE' }
  );
  return parseResponse<BudgetCategoryUsageDetail>(response);
}

export async function createCategory(
  profileId: number,
  payload: { name: string; color?: string },
  fetcher: typeof fetch = fetch
): Promise<BudgetCategory> {
  const response = await fetcher(`${API_BASE_URL}/budget/profiles/${profileId}/categories`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload)
  });
  return parseResponse<BudgetCategory>(response);
}

export async function updateCategory(
  profileId: number,
  categoryId: number,
  payload: {
    name?: string;
    color?: string;
    scope?: BudgetCategoryScope;
    year_month?: string;
  },
  fetcher: typeof fetch = fetch
): Promise<BudgetCategory> {
  const response = await fetcher(
    `${API_BASE_URL}/budget/profiles/${profileId}/categories/${categoryId}`,
    {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload)
    }
  );
  return parseResponse<BudgetCategory>(response);
}

export async function deleteCategory(
  profileId: number,
  categoryId: number,
  fetcher: typeof fetch = fetch
): Promise<void> {
  const response = await fetcher(
    `${API_BASE_URL}/budget/profiles/${profileId}/categories/${categoryId}`,
    { method: 'DELETE' }
  );
  await parseResponse<void>(response);
}

export async function listIncomeSources(
  profileId: number,
  fetcher: typeof fetch = fetch
): Promise<BudgetIncomeSource[]> {
  const response = await fetcher(`${API_BASE_URL}/budget/profiles/${profileId}/income-sources`);
  return parseResponse<BudgetIncomeSource[]>(response);
}

export async function createIncomeSource(
  profileId: number,
  payload: { name: string; recurrence_hint?: string },
  fetcher: typeof fetch = fetch
): Promise<BudgetIncomeSource> {
  const response = await fetcher(`${API_BASE_URL}/budget/profiles/${profileId}/income-sources`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload)
  });
  return parseResponse<BudgetIncomeSource>(response);
}

export async function deleteIncomeSource(
  profileId: number,
  sourceId: number,
  fetcher: typeof fetch = fetch
): Promise<void> {
  const response = await fetcher(`${API_BASE_URL}/budget/profiles/${profileId}/income-sources/${sourceId}`, {
    method: 'DELETE'
  });
  await parseResponse<void>(response);
}

export async function stopRecurringIncomeFromMonth(
  profileId: number,
  sourceId: number,
  fromYearMonth: string,
  fetcher: typeof fetch = fetch
): Promise<BudgetIncomeSource | null> {
  const response = await fetcher(
    `${API_BASE_URL}/budget/profiles/${profileId}/income-sources/${sourceId}/stop-from/${encodeURIComponent(fromYearMonth)}`,
    { method: 'POST' }
  );
  if (response.status === 204) {
    return null;
  }
  return parseResponse<BudgetIncomeSource>(response);
}

export async function createMonthExpense(
  profileId: number,
  yearMonth: string,
  payload: {
    description: string;
    event_date: string;
    amount_brl: number;
    category_id: number;
    tag_id?: number | null;
    recurring?: boolean;
    indefinite?: boolean;
    end_year_month?: string | null;
  },
  fetcher: typeof fetch = fetch
): Promise<BudgetMonthSnapshot> {
  const response = await fetcher(`${API_BASE_URL}/budget/profiles/${profileId}/months/${yearMonth}/expenses`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload)
  });
  return parseResponse<BudgetMonthSnapshot>(response);
}

export async function listRecurringExpenses(
  profileId: number,
  fetcher: typeof fetch = fetch
): Promise<BudgetRecurringExpense[]> {
  const response = await fetcher(`${API_BASE_URL}/budget/profiles/${profileId}/recurring-expenses`);
  return parseResponse<BudgetRecurringExpense[]>(response);
}

export async function updateRecurringExpense(
  profileId: number,
  ruleId: number,
  payload: {
    description?: string;
    event_date?: string;
    amount_brl?: number;
    category_id?: number;
    tag_id?: number | null;
    indefinite?: boolean;
    end_year_month?: string | null;
  },
  fetcher: typeof fetch = fetch
): Promise<BudgetRecurringExpense> {
  const response = await fetcher(
    `${API_BASE_URL}/budget/profiles/${profileId}/recurring-expenses/${ruleId}`,
    {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload)
    }
  );
  return parseResponse<BudgetRecurringExpense>(response);
}

export async function deleteRecurringExpense(
  profileId: number,
  ruleId: number,
  fetcher: typeof fetch = fetch
): Promise<void> {
  const response = await fetcher(`${API_BASE_URL}/budget/profiles/${profileId}/recurring-expenses/${ruleId}`, {
    method: 'DELETE'
  });
  await parseResponse<void>(response);
}

export async function stopRecurringExpenseFromMonth(
  profileId: number,
  ruleId: number,
  fromYearMonth: string,
  fetcher: typeof fetch = fetch
): Promise<BudgetRecurringExpense | null> {
  const response = await fetcher(
    `${API_BASE_URL}/budget/profiles/${profileId}/recurring-expenses/${ruleId}/stop-from/${encodeURIComponent(fromYearMonth)}`,
    { method: 'POST' }
  );
  if (response.status === 204) {
    return null;
  }
  return parseResponse<BudgetRecurringExpense>(response);
}

export async function getBudgetDashboard(
  profileId: number,
  focusYearMonth: string,
  options: {
    months?: 3 | 6;
    forwardMonths?: number;
    fromYearMonth?: string;
    toYearMonth?: string;
  } = {},
  fetcher: typeof fetch = fetch
): Promise<BudgetDashboard> {
  const params = new URLSearchParams({ focus: focusYearMonth });
  if (options.fromYearMonth && options.toYearMonth) {
    params.set('from', options.fromYearMonth);
    params.set('to', options.toYearMonth);
  } else {
    const months = options.months ?? 6;
    const forward = options.forwardMonths ?? months;
    params.set('months', String(months));
    params.set('forward', String(forward));
  }
  const response = await fetcher(
    `${API_BASE_URL}/budget/profiles/${profileId}/dashboard?${params.toString()}`
  );
  return parseResponse<BudgetDashboard>(response);
}
